embassy-stm32f4/src/serial.rs

```diff
@@ -0,0 +1,272 @@
+        }
+    }
+
+    /// Sends serial data.
+    ///
+    /// `tx_buffer` is marked as static as per `embedded-dma` requirements.
+    /// It it safe to use a buffer with a non static lifetime if memory is not
+    /// reused until the future has finished.
+    pub fn send<'a, B>(&'a mut self, tx_buffer: B) -> SendFuture<'a, B>
+    where
+        B: WriteBuffer<Word = u8> + 'static,
+    {
+        let tx_stream = self.tx_stream.take().unwrap();
+        let usart = self.usart.take().unwrap();
+        let mut tx_transfer = Transfer::init(
+            tx_stream,
+            usart,
+            tx_buffer,
+            None,
+            DmaConfig::default()
+                .transfer_complete_interrupt(true)
+                .memory_increment(true)
+                .double_buffer(false),
+        );
+
+        SendFuture {
+            uarte: self,
+            tx_transfer: Some(tx_transfer),
+            // tx_stream: Some(tx_stream),
+            // usart: Some(usart),
+        }
+    }
+
+    /// Receives serial data.
+    ///
+    /// The future is pending until the buffer is completely filled.
+    /// A common pattern is to use [`stop()`](ReceiveFuture::stop) to cancel
+    /// unfinished transfers after a timeout to prevent lockup when no more data
+    /// is incoming.
+    ///
+    /// `rx_buffer` is marked as static as per `embedded-dma` requirements.
+    /// It it safe to use a buffer with a non static lifetime if memory is not
+    /// reused until the future has finished.
+    pub fn receive<'a, B>(&'a mut self, rx_buffer: B) -> ReceiveFuture<'a, B>
+    where
+        B: WriteBuffer<Word = u8> + 'static,
+    {
+        let rx_stream = self.rx_stream.take().unwrap();
+        let usart = self.usart.take().unwrap();
+        let mut rx_transfer = Transfer::init(
+            rx_stream,
+            usart,
+            rx_buffer,
+            None,
+            DmaConfig::default()
+                .transfer_complete_interrupt(true)
+                .half_transfer_interrupt(true)
+                .memory_increment(true)
+                .double_buffer(false),
+        );
+
+        ReceiveFuture {
+            uarte: self,
+            rx_transfer: Some(rx_transfer),
+        }
+    }
+}
+
+/// Future for the [`LowPowerUarte::send()`] method.
+pub struct SendFuture<'a, B: WriteBuffer<Word = u8> + 'static> {
+    uarte: &'a mut Uarte,
+    tx_transfer: Option<Transfer<Stream7<DMA2>, Channel4, USART1, MemoryToPeripheral, B>>,
+}
+
+impl<'a, B> Drop for SendFuture<'a, B>
+where
+    B: WriteBuffer<Word = u8> + 'static,
+{
+    fn drop(self: &mut Self) {}
+}
+
+impl<'a, B> Future for SendFuture<'a, B>
+where
+    B: WriteBuffer<Word = u8> + 'static,
+{
+    type Output = ();
+
+    fn poll(self: core::pin::Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<()> {
+        let Self { uarte, tx_transfer } = unsafe { self.get_unchecked_mut() };
+
+        if true {
+            // tx_transfer.unwrap().is_done() {
```

Comment:
I can't call this method, because the borrow checker complains. Need a solution to this.